File: JsonSchemaAndStructs/KiteConnect-Dart-Structs/margin_commodity.dart
// To parse this JSON data, do
//
//     final marginCommodity = marginCommodityFromMap(jsonString);

import 'package:freezed_annotation/freezed_annotation.dart';
import 'dart:convert';

part 'margin_commodity.freezed.dart';
part 'margin_commodity.g.dart';

@freezed
abstract class MarginCommodity with _$MarginCommodity {
    const factory MarginCommodity({
        Data data,
        String status,
    }) = _MarginCommodity;

    factory MarginCommodity.fromJson(Map<String, dynamic> json) => _$MarginCommodityFromJson(json);
}

@freezed
abstract class Data with _$Data {
    const factory Data({
        Available available,
        bool enabled,
        double net,
        Map<String, double> utilised,
    }) = _Data;

    factory Data.fromJson(Map<String, dynamic> json) => _$DataFromJson(json);
}

@freezed
abstract class Available with _$Available {
    const factory Available({
        int adhocMargin,
        double cash,
        int collateral,
        int intradayPayin,
        double liveBalance,
        double openingBalance,
    }) = _Available;

    factory Available.fromJson(Map<String, dynamic> json) => _$AvailableFromJson(json);
}

File: JsonSchemaAndStructs/KiteConnect-Haskell-Structs/margin_commodity.hs
{-# LANGUAGE StrictData #-}
{-# LANGUAGE OverloadedStrings #-}

module MarginCommodity
    ( MarginCommodity (..)
    , Data (..)
    , Available (..)
    , decodeTopLevel
    ) where

import Data.Aeson
import Data.Aeson.Types (emptyObject)
import Data.ByteString.Lazy (ByteString)
import Data.HashMap.Strict (HashMap)
import Data.Text (Text)
import Data.Vector (Vector)

data MarginCommodity = MarginCommodity
    { marginCommodityDataMarginCommodity :: Maybe Data
    , statusMarginCommodity :: Maybe Text
    } deriving (Show)

data Data = Data
    { availableData :: Maybe Available
    , enabledData :: Maybe Bool
    , netData :: Maybe Float
    , utilisedData :: Maybe (HashMap Text Float)
    } deriving (Show)

data Available = Available
    { adhocMarginAvailable :: Maybe Int
    , cashAvailable :: Maybe Float
    , collateralAvailable :: Maybe Int
    , intradayPayinAvailable :: Maybe Int
    , liveBalanceAvailable :: Maybe Float
    , openingBalanceAvailable :: Maybe Float
    } deriving (Show)

decodeTopLevel :: ByteString -> Maybe MarginCommodity
decodeTopLevel = decode

instance ToJSON MarginCommodity where
    toJSON (MarginCommodity marginCommodityDataMarginCommodity statusMarginCommodity) =
        object
        [ "data" .= marginCommodityDataMarginCommodity
        , "status" .= statusMarginCommodity
        ]

instance FromJSON MarginCommodity where
    parseJSON (Object v) = MarginCommodity
        <$> v .:? "data"
        <*> v .:? "status"

instance ToJSON Data where
    toJSON (Data availableData enabledData netData utilisedData) =
        object
        [ "available" .= availableData
        , "enabled" .= enabledData
        , "net" .= netData
        , "utilised" .= utilisedData
        ]

instance FromJSON Data where
    parseJSON (Object v) = Data
        <$> v .:? "available"
        <*> v .:? "enabled"
        <*> v .:? "net"
        <*> v .:? "utilised"

instance ToJSON Available where
    toJSON (Available adhocMarginAvailable cashAvailable collateralAvailable intradayPayinAvailable liveBalanceAvailable openingBalanceAvailable) =
        object
        [ "adhoc_margin" .= adhocMarginAvailable
        , "cash" .= cashAvailable
        , "collateral" .= collateralAvailable
        , "intraday_payin" .= intradayPayinAvailable
        , "live_balance" .= liveBalanceAvailable
        , "opening_balance" .= openingBalanceAvailable
        ]

instance FromJSON Available where
    parseJSON (Object v) = Available
        <$> v .:? "adhoc_margin"
        <*> v .:? "cash"
        <*> v .:? "collateral"
        <*> v .:? "intraday_payin"
        <*> v .:? "live_balance"
        <*> v .:? "opening_balance"
